File: Windows/src/AudioLocal.Windows.Core/Services/WavStitcher.cs
using AudioLocal.Windows.Core.Models;

namespace AudioLocal.Windows.Core.Services;

public sealed class WavStitcher
{
    public StitchedAudio Stitch(IReadOnlyList<ChapterWavSegment> chapters, double pauseBetweenChaptersSeconds = 0.85)
    {
        if (chapters.Count == 0)
        {
            throw new InvalidOperationException("No chapter audio was generated.");
        }

        var parsed = chapters.Select(static chapter => (chapter.Title, Segment: ParseWav(chapter.WavData))).ToArray();
        var first = parsed[0].Segment;

        if (parsed.Any(segment => segment.Segment.Format != first.Format))
        {
            throw new InvalidOperationException("The generated chapter audio uses mismatched WAV formats.");
        }

        var silenceFrameCount = (int)Math.Round(first.Format.SampleRate * pauseBetweenChaptersSeconds, MidpointRounding.AwayFromZero);
        var silenceByteCount = silenceFrameCount * first.Format.BlockAlign;
        var silence = new byte[Math.Max(0, silenceByteCount)];

        var markers = new List<ChapterMarker>(parsed.Length);
        using var pcmBuffer = new MemoryStream();
        var cursorMs = 0;

        for (var index = 0; index < parsed.Length; index++)
        {
            var (title, segment) = parsed[index];
            var durationMs = segment.DurationMilliseconds();
            markers.Add(new ChapterMarker(title, cursorMs, cursorMs + durationMs));
            pcmBuffer.Write(segment.PcmData);

            cursorMs += durationMs;
            if (index < parsed.Length - 1 && silence.Length > 0)
            {
                pcmBuffer.Write(silence);
                cursorMs += (int)Math.Round(pauseBetweenChaptersSeconds * 1000, MidpointRounding.AwayFromZero);
            }
        }

        return new StitchedAudio(
            WavEncoder.MakeWav(pcmBuffer.ToArray(), first.Format.SampleRate, first.Format.Channels, first.Format.BitsPerSample),
            markers);
    }

    private static ParsedWav ParseWav(byte[] data)
    {
        if (data.Length < 44 ||
            data[0] != (byte)'R' || data[1] != (byte)'I' || data[2] != (byte)'F' || data[3] != (byte)'F' ||
            data[8] != (byte)'W' || data[9] != (byte)'A' || data[10] != (byte)'V' || data[11] != (byte)'E')
        {
            throw new InvalidOperationException("The generated chapter audio is not PCM WAV.");
        }

        var offset = 12;
        WavFormat? format = null;
        byte[]? pcmData = null;

        while (offset + 8 <= data.Length)
        {
            var chunkId = System.Text.Encoding.ASCII.GetString(data, offset, 4);
            var chunkSize = BitConverter.ToInt32(data, offset + 4);
            var chunkStart = offset + 8;
            var nextChunkOffset = chunkStart + chunkSize + (chunkSize % 2);

            if (nextChunkOffset > data.Length)
            {
                throw new InvalidOperationException("The generated chapter audio is malformed.");
            }

            if (chunkId == "fmt ")
            {
                var audioFormat = BitConverter.ToInt16(data, chunkStart);
                var channels = BitConverter.ToInt16(data, chunkStart + 2);
                var sampleRate = BitConverter.ToInt32(data, chunkStart + 4);
                var blockAlign = BitConverter.ToInt16(data, chunkStart + 12);
                var bitsPerSample = BitConverter.ToInt16(data, chunkStart + 14);

                if (audioFormat != 1 || bitsPerSample != 16)
                {
                    throw new InvalidOperationException("The generated chapter audio must be 16-bit PCM WAV.");
                }

                format = new WavFormat(sampleRate, channels, bitsPerSample, blockAlign);
            }
            else if (chunkId == "data")
            {
                pcmData = data.Skip(chunkStart).Take(chunkSize).ToArray();
            }

            offset = nextChunkOffset;
        }

        if (format is null || pcmData is null)
        {
            throw new InvalidOperationException("The generated chapter audio could not be stitched.");
        }

        return new ParsedWav(format, pcmData);
    }

    private sealed record ParsedWav(WavFormat Format, byte[] PcmData)
    {
        public int DurationMilliseconds() =>
            (int)Math.Round((double)PcmData.Length / Format.BlockAlign / Format.SampleRate * 1000, MidpointRounding.AwayFromZero);
    }

    private sealed record WavFormat(int SampleRate, short Channels, short BitsPerSample, short BlockAlign);
}

File: Sources/AudioLocal/Services/WAVStitcher.swift
import Foundation

struct WAVStitcher: Sendable {
    enum StitchError: LocalizedError {
        case noAudio
        case unsupportedFormat
        case mismatchedFormats

        var errorDescription: String? {
            switch self {
            case .noAudio:
                return "No chapter audio was generated."
            case .unsupportedFormat:
                return "The generated chapter audio could not be stitched because it is not PCM WAV."
            case .mismatchedFormats:
                return "The generated chapter audio could not be stitched because the WAV formats do not match."
            }
        }
    }

    func stitch(chapterAudio: [Data], pauseBetweenChapters: TimeInterval = 0.85) throws -> Data {
        guard !chapterAudio.isEmpty else {
            throw StitchError.noAudio
        }

        let parsedSegments = try chapterAudio.map(parseWAV)
        guard let firstSegment = parsedSegments.first else {
            throw StitchError.noAudio
        }

        for segment in parsedSegments.dropFirst() where segment.format != firstSegment.format {
            throw StitchError.mismatchedFormats
        }

        if parsedSegments.count == 1 {
            return chapterAudio[0]
        }

        let silenceFrameCount = Int((Double(firstSegment.format.sampleRate) * pauseBetweenChapters).rounded())
        let silenceByteCount = silenceFrameCount * Int(firstSegment.format.blockAlign)
        let silence = Data(repeating: 0, count: max(silenceByteCount, 0))

        var combinedPCM = Data()
        for (index, segment) in parsedSegments.enumerated() {
            combinedPCM.append(segment.pcmData)
            if index < parsedSegments.count - 1 {
                combinedPCM.append(silence)
            }
        }

        return WAVEncoder.makeWAV(
            pcmData: combinedPCM,
            sampleRate: Int(firstSegment.format.sampleRate),
            channels: Int(firstSegment.format.channels),
            bitsPerSample: Int(firstSegment.format.bitsPerSample)
        )
    }

    private func parseWAV(_ data: Data) throws -> ParsedWAV {
        guard data.count >= 44 else {
            throw StitchError.unsupportedFormat
        }

        guard data.asciiString(at: 0, count: 4) == "RIFF",
              data.asciiString(at: 8, count: 4) == "WAVE" else {
            throw StitchError.unsupportedFormat
        }

        var offset = 12
        var format: WAVFormat?
        var pcmData: Data?

        while offset + 8 <= data.count {
            let chunkID = data.asciiString(at: offset, count: 4)
            let chunkSize = Int(data.readUInt32LE(at: offset + 4))
            let chunkStart = offset + 8
            let nextChunkOffset = chunkStart + chunkSize + (chunkSize % 2)

            guard nextChunkOffset <= data.count else {
                throw StitchError.unsupportedFormat
            }

            if chunkID == "fmt " {
                guard chunkSize >= 16 else {
                    throw StitchError.unsupportedFormat
                }

                let audioFormat = data.readUInt16LE(at: chunkStart)
                let channels = data.readUInt16LE(at: chunkStart + 2)
                let sampleRate = data.readUInt32LE(at: chunkStart + 4)
                let blockAlign = data.readUInt16LE(at: chunkStart + 12)
                let bitsPerSample = data.readUInt16LE(at: chunkStart + 14)

                guard audioFormat == 1, bitsPerSample == 16 else {
                    throw StitchError.unsupportedFormat
                }

                format = WAVFormat(
                    sampleRate: sampleRate,
                    channels: channels,
                    bitsPerSample: bitsPerSample,
                    blockAlign: blockAlign
                )
            } else if chunkID == "data" {
                pcmData = data.subdata(in: chunkStart..<chunkStart + chunkSize)
            }

            offset = nextChunkOffset
        }

        guard let format, let pcmData else {
            throw StitchError.unsupportedFormat
        }

        return ParsedWAV(format: format, pcmData: pcmData)
    }
}

private struct ParsedWAV {
    let format: WAVFormat
    let pcmData: Data
}

private struct WAVFormat: Equatable {
    let sampleRate: UInt32
    let channels: UInt16
    let bitsPerSample: UInt16
    let blockAlign: UInt16
}

private extension Data {
    func asciiString(at offset: Int, count: Int) -> String {
        String(decoding: self[offset..<offset + count], as: UTF8.self)
    }

    func readUInt16LE(at offset: Int) -> UInt16 {
        let lower = UInt16(self[offset])
        let upper = UInt16(self[offset + 1]) << 8
        return lower | upper
    }

    func readUInt32LE(at offset: Int) -> UInt32 {
        let b0 = UInt32(self[offset])
        let b1 = UInt32(self[offset + 1]) << 8
        let b2 = UInt32(self[offset + 2]) << 16
        let b3 = UInt32(self[offset + 3]) << 24
        return b0 | b1 | b2 | b3
    }
}
